### PR TITLE
[loop] moved loop iteration utils to LoopUtils, so that Middle end pa…

### PR DESCRIPTION
**Pull request by F-Stuckmann**
moved loop iteration utils to LoopUtils, so that Middle end passes also have access

### CODE REVIEW

File: llvm/include/llvm/Transforms/Utils/LoopUtils.h [hunk 1]
```diff
@@ -283,6 +284,22 @@ MDNode *updateIterCounts(LLVMContext &Context, MDNode *LoopID,
...
+
+/// Get the LoopID from a single block loop or nullptr
+const MDNode *getLoopID(const MachineBasicBlock &LoopBlock);
+
+/// Get LoopID for Latch
+const MDNode *getLoopID(const BasicBlock *LoopBB);
+
+/// LoopBlock should hold the backedge. Since we generally call it for
+/// single block loops, that's automatically true.
+std::optional<int64_t> getMinTripCount(const MachineBasicBlock &LoopBlock);
```
Comment (gbossu) [inline llvm/include/llvm/Transforms/Utils/LoopUtils.h]: Could you leave helper functions that deal with `MachineBasicBlock` out of this header? I believe it's better to keep LoopInfo about LLVM IR, not MIR.

File: llvm/lib/Transforms/Utils/LoopUtils.cpp [hunk 2]
```diff
@@ -415,6 +417,87 @@ MDNode *llvm::updateIterCounts(LLVMContext &Context, MDNode *LoopID,
...
+const MDNode *llvm::getLoopID(const BasicBlock *LoopBB) {
+  if (!LoopBB)
+    return nullptr;
+  const Instruction *TI = LoopBB->getTerminator();
+  if (!TI)
+    return nullptr;
+
+  const MDNode *LoopID = TI->getMetadata(LLVMContext::MD_loop);
+  return LoopID;
+}
```
Comment (gbossu) [inline llvm/lib/Transforms/Utils/LoopUtils.cpp]: This seems to be a new function. Is that different from `LoopT *LoopInfoBase::getLoopFor(const BlockT *BB)`? I think we should avoid adding such low-level helpers if there's already a similar one.

File: llvm/include/llvm/Transforms/Utils/LoopUtils.h [hunk 3]
```diff
@@ -283,6 +284,22 @@ MDNode *updateIterCounts(LLVMContext &Context, MDNode *LoopID,
...
+
```
Comment (andcarminati) [inline llvm/include/llvm/Transforms/Utils/LoopUtils.h]: I think we can handle everything here using `class BasicBlock`. In the backend we can have a wrapper function that uses *[getBasicBlock](https://llvm.org/doxygen/classllvm_1_1MachineBasicBlock.html#a4874816314c3308be0bf1e71de2078d8)(). I agree that this is the place for IR-only code.

File: llvm/include/llvm/Transforms/Utils/LoopUtils.h [hunk 4]
```diff
@@ -283,6 +284,22 @@ MDNode *updateIterCounts(LLVMContext &Context, MDNode *LoopID,
...
+/// Loop for the attribute that provides minimum iteration counts
```
Comment (martien-de-jong) [inline llvm/include/llvm/Transforms/Utils/LoopUtils.h]: *Look*

File: llvm/lib/Transforms/Utils/LoopUtils.cpp [hunk 5]
```diff
@@ -415,6 +417,87 @@ MDNode *llvm::updateIterCounts(LLVMContext &Context, MDNode *LoopID,
...
+  assert(dyn_cast<MDString>(LoopID->getOperand(0))->getString() ==
+         LLVMLoopIterCount);
+  if (LoopID) {
+    assert((LoopID->getNumOperands() >= 2 && LoopID->getNumOperands() <= 3) &&
+           "Iteration count hint should have one or two numeric operands.");
+    int64_t MinTripCount =
+        mdconst::extract<ConstantInt>(LoopID->getOperand(1))->getSExtValue();
+    assert(MinTripCount >= 0 && "Range lwb should not be negative.");
+    // LLVM_DEBUG(dbgs() << "AIELoopUtils: MinTripCount from pragma =  "
+    //                   << MinTripCount << "\n");
```
Comment (martien-de-jong) [inline llvm/lib/Transforms/Utils/LoopUtils.cpp]: I'd say remove this LLVM_DEBUG, but at least get rid of AIE

File: llvm/include/llvm/Transforms/Utils/LoopUtils.h (hunk 4 above)
Comment (F-Stuckmann) [reply llvm/include/llvm/Transforms/Utils/LoopUtils.h]: fixed it

File: llvm/lib/Transforms/Utils/LoopUtils.cpp (hunk 5 above)
Comment (F-Stuckmann) [reply llvm/lib/Transforms/Utils/LoopUtils.cpp]: fixed it

File: llvm/include/llvm/Transforms/Utils/LoopUtils.h (hunk 3 above)
Comment (F-Stuckmann) [reply llvm/include/llvm/Transforms/Utils/LoopUtils.h]: removed the wrapper in `LoopUtils.h` and moved it to `AIELoopUtils.h`

File: llvm/lib/Transforms/Utils/LoopUtils.cpp (hunk 2 above)
Comment (F-Stuckmann) [reply llvm/lib/Transforms/Utils/LoopUtils.cpp]: I left this function alone in `AIELoopUtils.cpp`

File: llvm/include/llvm/Transforms/Utils/LoopUtils.h (hunk 1 above)
Comment (F-Stuckmann) [reply llvm/include/llvm/Transforms/Utils/LoopUtils.h]: removed the `MachineBasicBlock `from `LoopUtils`

**Review by gbossu — APPROVED** [review]
LGTM

**Pull request merged**